redisstorage.go

```diff
@@ -65,47 +86,78 @@ func (s *Storage) GetCookieJar() http.CookieJar {
 // SetCookies implements http/CookieJar.SetCookies()
 func (s *Storage) SetCookies(u *url.URL, cookies []*http.Cookie) {
 	// TODO RFC 6265
-	cookieStrings := make([]string, 0, len(cookies))
-	for _, c := range cookies {
-		cookieStrings = append(cookieStrings, c.String())
+
+	// TODO(js) Cookie methods currently have no way to return an error.
+
+	// We need to use a write lock to prevent a race in the db:
+	// if two callers set cookies in a very small window of time,
+	// it is possible to drop the new cookies from one caller
+	// ('last update wins' == best avoided).
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	cookieStr, err := s.Client.Get(s.getCookieID(u.Host)).Result()
+	if err == redis.Nil {
+		cookieStr = ""
+	} else if err != nil {
+		// return nil
+		log.Printf("SetCookies() .Get error %s", err)
+		return
 	}
-	for _, c := range s.Cookies(u) {
-		duplication := false
-		for _, c2 := range cookies {
-			if c2.Name == c.Name {
-				duplication = true
-				break
-			}
-		}
-		if !duplication {
-			cookieStrings = append(cookieStrings, c.String())
+
+	// Merge existing cookies, new cookies have precendence.
+	cnew := make([]*http.Cookie, len(cookies))
+	copy(cnew, cookies)
+	existing := unstringify(cookieStr)
+	for _, c := range existing {
+		if !contains(cnew, c.Name) {
+			cnew = append(cnew, c)
 		}
 	}
```

Comment:
The previous version of the above was giving higher precedence to existing cookies, preventing new cookie values from being set.